iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl

```diff
@@ -65,122 +92,146 @@ inline bool SpscSofi<ValueType, CapacityValue>::setCapacity(const uint64_t newSi
     } while (!m_readPosition.compare_exchange_weak(
-        currentReadPosition, nextReadPosition, std::memory_order_acq_rel, std::memory_order_acquire));
+        currentReadPos, currentReadPos + 1U, std::memory_order_relaxed, std::memory_order_relaxed));
 
-    return popWasSuccessful;
+    return true;
 }
 
 template <class ValueType, uint64_t CapacityValue>
 inline bool SpscSofi<ValueType, CapacityValue>::push(const ValueType& valueIn, ValueType& valueOut) noexcept
 {
     constexpr bool SOFI_OVERFLOW{false};
 
-    uint64_t currentWritePosition = m_writePosition.load(std::memory_order_relaxed);
-    uint64_t nextWritePosition = currentWritePosition + 1U;
-
-    m_data[currentWritePosition % m_size] = valueIn;
-    m_writePosition.store(nextWritePosition, std::memory_order_release);
-
-    uint64_t currentReadPosition = m_readPosition.load(std::memory_order_acquire);
-
-    // check if there is a free position for the next push
-    if (nextWritePosition < currentReadPosition + m_size)
+    // SYNC POINT READ: m_data
+    // We need to synchronize data to avoid the following scenario:
+    // 1. A thread calls push() and updates data with a new value
+    // 2. Another thread calls push() and enters the overflow case. The data could be read before
+    // any synchronization
+    uint64_t currentWritePos = m_writePosition.load(std::memory_order_acquire);
+    uint64_t nextWritePos = currentWritePos + 1U;
+
+    m_data[currentWritePos % m_size] = valueIn;
+    // SYNC POINT WRITE: m_data
+    // We need to make sure that writing the value happens before incrementing the
+    // m_writePosition otherwise the following scenario can happen:
+    // 1. m_writePosition is increased (but the value has not been written yet)
+    // 2. Another thread calls pop(): we check if the queue is empty => no (e.g. m_writePosition == 1
+    // and m_readPosition == 0)
+    // 3. In pop(), a data race can occur
+    // With memory_order_release, this cannot happen as it is guaranteed that writing the data
+    // happens before incrementing m_writePosition
+    // =======================================
+    // Note that the following situation can still happen (but is not a problem):
+    // 1. A value is written (m_writePosition hasn't been incremented yet)
+    // 2. Another thread calls pop(): we check if the queue is empty => yes (e.g. m_writePosition ==
+    // m_readPosition == 0 )
+    // 3. An element was already stored so we could have popped the element
+    m_writePosition.store(nextWritePos, std::memory_order_release);
+
+    // Memory order relaxed is enough since:
+    // - synchronization is not needed with m_readPosition
+    // - operation reordering:
+    //    - cannot move below if statement otherwise, the code won't compile
+    //    - if it moves above, we might get an outdated read position that will be caught by the
+    //    compare_exchange check
+    uint64_t currentReadPos = m_readPosition.load(std::memory_order_relaxed);
```

Comment:
This needs to be reverted since it acts as sync point to the CAS loop in the pop method.